### ISSUE TITLE
rustls ech-client example DH KEM support

### ISSUE THREAD

**Issue by cpu** [defo-project/ech-interop-report#2]
Hi :wave:

I just spotted this text in `interop.tex`:

https://github.com/defo-project/ech-interop-report/blob/fce391f02e6e4dcefa94760db9ef798a4214cfa8/source/interop.tex#L193-L201

I think (and I suspect you're aware now too) this is actually https://github.com/rustls/rustls/issues/2276 and not related to the curve choice.

> as that client only supports x25519 (presumably)

FWIW the default HPKE backend used by the `ech-client.rs` example supports [a range of suites](https://github.com/rustls/rustls/blob/9045f1044166f7170df2c58e7f4d3447c84d9ff0/rustls/src/crypto/aws_lc_rs/hpke.rs#L26-L46) that includes p-256, p-384 and p-521 DH KEMs in addition to x25519.

I haven't done an end-to-end read of your report yet (thank you btw! very valuable effort) but would be interested in digging in deeper if it turns out the failures observed weren't from https://github.com/rustls/rustls/issues/2276 Please feel free to tag me somewhere with details.

Thanks!

**Comment by sftcd** [comment]
Thanks for that, you're right that https://github.com/rustls/rustls/issues/2276 is the relevant thing here and as that's described below, so I just deleted the text you quoted above.

Be very happy to get any more comments.

**Comment by sftcd** [comment]
A question back to you though:

> FWIW the default HPKE backend used by the ech-client.rs example supports [a range of suites](https://github.com/rustls/rustls/blob/9045f1044166f7170df2c58e7f4d3447c84d9ff0/rustls/src/crypto/aws_lc_rs/hpke.rs#L26-L46) that includes p-256, p-384 and p-521 DH KEMs in addition to x25519.

I still see p256 as not working for ECH though, so is it correct to say that only x25519 is supported for ECH, even if more KEMs are supported in HPKE?

```
$ dig +short https p256-ng.test.defo.ie
1 . ipv4hint=185.88.140.5 ech=AGf+DQBjngAQAEEEFgkLkdHS+tz18PXmdYpYBZdKGxjBm+WNG9AsKNaOCj4bmDbuF4KyvHEZGj6+Wls1VBe1OXierpjGiYOEriOWnAAEAAIAAwATcHVibGljLnRlc3QuZGVmby5pZQAA ipv6hint=2a00:c6c0:0:134:2::1
$ ./target/debug/ech-rs-client --path "echstat.php?format=json" public.test.defo.ie p256-ng.test.defo.ie 
thread 'main' panicked at src/main.rs:61:14:
called `Result::unwrap()` on an `Err` value: InvalidEncryptedClientHello(NoCompatibleConfig)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Thanks

**Comment by cpu** [comment]
> I still see p256 as not working for ECH though, so is it correct to say that only x25519 is supported for ECH, even if more KEMs are supported in HPKE?

No, that shouldn't be correct. 

I will investigate (likely early next week) and get back to you. Thanks!

**Comment by sftcd** [comment]
Great, ping if I can help testing something.

**Comment by ctz** [comment]
Apologies for butting in, but I got here via the TLSWG post :)

It seems this case doesn't work because it demands:

```
ech_configs [
    V18(
        EchConfigContents {
            key_config: HpkeKeyConfig {
                config_id: 158,
                kem_id: DHKEM_P256_HKDF_SHA256,
                public_key: 0416090b91d1d2fadcf5f0f5e6758a5805974a1b18c19be58d1bd02c28d68e0a3e1b9836ee1782b2bc71191a3ebe5a5b355417b539789eae98c6898384ae23969c,
                symmetric_cipher_suites: [
                    HpkeSymmetricCipherSuite {
                        kdf_id: HKDF_SHA384,
                        aead_id: CHACHA20_POLY_1305,
                    },
                ],
            },
            maximum_name_length: 0,
            public_name: DnsName(
                "public.test.defo.ie",
            ),
            extensions: [],
        },
    ),
]
```

We don't have a matching `HpkeSuite` for that, the closest we get is:

```
    HpkeSuite {
        kem: DHKEM_P384_HKDF_SHA384,
        sym: HpkeSymmetricCipherSuite {
            kdf_id: HKDF_SHA384,
            aead_id: CHACHA20_POLY_1305,
        },
    },
    HpkeSuite {
        kem: DHKEM_P256_HKDF_SHA256,
        sym: HpkeSymmetricCipherSuite {
            kdf_id: HKDF_SHA256,
            aead_id: CHACHA20_POLY_1305,
        },
    },
```

I guess my question is: do people _really_ want to have full mix-and-match of HPKE suite components like this, rather than roughly-matching combinations?

**Comment by cpu** [comment]
> Apologies for butting in, but I got here via the TLSWG post :)

Thanks for taking a look! I had this on my list for this week but hadn't gotten to it yet. Your conclusion matches what I expected to find for whatever that's worth :-)

> do people really want to have full mix-and-match of HPKE suite components like this, rather than roughly-matching combinations?

It doesn't feel particularly worthwhile to me to support the full set of all possible combinations. I doubt browser support is that extensive and I think that's likely to be the main motivator for the rest of the ecosystem?

**Comment by sftcd** [comment]
> Apologies for butting in, but I got here via the TLSWG post :)

You're most welcome to chime in! We wanna get this as accurate as we can.

So from the above I guess rustls supports p384 (with sha384). Choosing that hash with that group makes sense of course, but why that group? Is it also supported in browsers or something? 

I totally appreciate that the full set of 96 crypto combos defined in RFC9180 is silly, other than for a (masochistic;-) library like OpenSSL, but not sure how to identify a good (test) subset. In the absence of that, I just added a p256 test case to see what'd happen.

I'm fine with adding another p384 test case if that's useful of course. (Be a few days maybe.)

**Comment by sftcd** [comment]
Oh, sorry, now I read more closely it looks like my chosen combo was a bit oddball - but do you need an exact match on all of the HPKE suite? I'd have thought each of the KEM/KDF/AEAD would be separately supported or not?

In any case I should clarify the oddball-ness of my choice in the doc.

**Comment by ctz** [comment]
Yeah at the minute we need an exact match for each component. And... I think that is moderately common in HPKE implementations?

eg, https://github.com/ctz/hpke-py/blob/main/hpke/hpke.py#L621-L628 (ok, that doesn't really count since I wrote it, but eh.)

https://docs.rs/hpke/latest/hpke/ (this one I didn't write, and is not the implementation we use, but requires the combination to be resolved _at compile time_ so will never be able to support all combinations unless every single one is written out, with code monomorphized for each one.)

It's a real shame ECH nor RFC9180 don't bless any MTI combinations.

**Comment by sftcd** [comment]
So I guess that makes my choice a good and a bad test case:-) I've added a note to add a more likely to be widely supported suite and will add some text based on the the above too. 

And yes, an MTI choice would have been better than 96 options for sure. I think that's an argument I lost (there was more than one, so I could be misremembering:-)

**Comment by sftcd** [comment]
I added some text in https://github.com/defo-project/ech-interop-report/commit/703eaa92d6f6271a64913edb014251c91fe6e62d - feel free to improve if needed and I'll add that new test case later.

**Comment by sftcd** [comment]
> I think that is moderately common in HPKE implementations?

OpenSSL code (from me:-) doesn't do that, but I'm not sure in general as most implementations probably mostly tested HPKE suites based on the test vectors in 9180 so that'd be less likely to show up.

**Comment by sftcd** [comment]
closing this as I'm starting work on an update to the doc and the oddball nature of some of the tests was noted in the January version.
